pkg/syncer/reconcile/fight/detector.go

```diff
@@ -73,15 +73,17 @@ func NewDetector() Detector {
 }
 
 // DetectFight detects whether the resource is needing updates too frequently.
-// If so, it increments the resource_fights metric and logs to klog.Warning.
-func (d *Detector) DetectFight(ctx ctx.Context, time time.Time, obj *unstructured.Unstructured, operation string) bool {
-	if fight := d.markUpdated(time, obj); fight != nil {
+// If so, it increments the resource_fights metric and logs to klog.Error.
+func (d *Detector) DetectFight(ctx ctx.Context, time time.Time,
+	obj *unstructured.Unstructured, operation string) (bool, status.ResourceError) {
+	fightErr := d.markUpdated(time, obj)
```

Comment:
This is a little weird. The `markUpdated` method should return "mark errors" or "update errors". Calling them "fight errors" leaks implementation details of the function not implied by the name.

And if you change this to just `err` then it's confusing why you're passing all errors to the `logFight` function.

Maybe this function should instead do an error type check for FightError and only call logFight if it's confirmed to be a FightError?

Otherwise, maybe there's a way to combine these functions? Or were they made separate on purpose to scope the lock to not include RecordResourceFight &  logFight?

Reply:
We don't need a type check for FightError because this method only detects the fight error, but not other types.

I think the confusion might be caused by the function name. I refactored it a little bit. Hopefully it is clearer than before.